staging/src/k8s.io/apiextensions-apiserver/third_party/forked/celopenapi/model/schemas.go

```diff
@@ -0,0 +1,125 @@
+	root := SchemaDeclType(s).MaybeAssignTypeName(maybeRootType)
+	types := FieldTypeMap(maybeRootType, root)
+	return root, types
+}
+
+// SchemaDeclType returns the cel type name associated with the schema element.
+func SchemaDeclType(s *schema.Structural) *DeclType {
+	if s.XIntOrString { // this extension is not required to have a type, so it must be handled before type lookup
+		return intOrStringType
+	}
+
+	declType, found := openAPISchemaTypes[s.Type]
+	if !found {
+		return nil
+	}
+
+	// We ignore XPreserveUnknownFields since we don't support validation rules on
+	// data that we don't have schema information for.
+
+	// We don't special case XEmbeddedResource. XEmbeddedResource identifies a field that contains an
+	// embedded kubernetes resource, but the contents of any validated fields must
+	// still be declared in the schema. If 'metadata' is declared explicitly, it may be
+	// referenced in validator rules. If other properties are declared explicitly, they may
+	// be referenced in validator rules. The above comment about XPreserveUnknownFields
+	// applies.
+
+	switch declType.TypeName() {
+	case ListType.TypeName():
+		return NewListType(SchemaDeclType(s.Items))
+	case MapType.TypeName():
+		if s.AdditionalProperties != nil && s.AdditionalProperties.Structural != nil {
+			return NewMapType(StringType, SchemaDeclType(s.AdditionalProperties.Structural))
+		}
+		fields := make(map[string]*DeclField, len(s.Properties))
+
+		required := map[string]bool{}
+		if s.ValueValidation != nil {
+			for _, f := range s.ValueValidation.Required {
+				required[f] = true
+			}
+		}
+		for name, prop := range s.Properties {
+			var enumValues []interface{}
+			if prop.ValueValidation != nil {
+				for _, e := range prop.ValueValidation.Enum {
+					enumValues = append(enumValues, e.Object)
+				}
+			}
+			if fieldType := SchemaDeclType(&prop); fieldType != nil {
+				fields[Escape(name)] = &DeclField{
+					Name:         Escape(name),
+					Required:     required[name],
+					Type:         fieldType,
+					defaultValue: prop.Default.Object,
+					enumValues:   enumValues, // Enum values are represented as strings in CEL
+				}
+			}
+		}
+		return NewObjectType("object", fields)
+	case StringType.TypeName():
+		if s.ValueValidation != nil {
+			switch s.ValueValidation.Format {
+			case "byte":
+				return StringType // OpenAPIv3 byte format represents base64 encoded string
+			case "binary":
+				return BytesType
+			case "duration":
+				return DurationType
+			case "date", "date-time":
+				return TimestampType
+			}
+		}
+	}
+	return declType
+}
+
+var (
+	openAPISchemaTypes = map[string]*DeclType{
+		"boolean":         BoolType,
+		"number":          DoubleType,
+		"integer":         IntType,
+		"null":            NullType,
+		"string":          StringType,
+		"date":            DateType,
+		"array":           ListType,
+		"object":          MapType,
+	}
```

Comment:
date, null are no OpenAPI v3 types. Something is off here. Dates are specified by the format, null is just a boolean property of a field called `nullable`.

Reply:
`date` is my fault, I mis-interpreted the list of types defined in https://kubernetes.io/docs/tasks/extend-kubernetes/custom-resources/custom-resource-definitions/ for print columns.  I'll remove it.

`null` is preexisting in this code, but looks we don't need it either, so I'll remove it too.